XMLHttpRequestProgressEventThrottle can be a weak pointer without being refable

youennf opened the pull request:
#### 0bce4652d6d810addfc0f462796a4f6683e982e2
<pre>
XMLHttpRequestProgressEventThrottle can be a weak pointer without being refable
<a href="https://rdar.apple.com/155610120">rdar://155610120</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295774">https://bugs.webkit.org/show_bug.cgi?id=295774</a>

Reviewed by Anne Van Kesteren.

We no longer need to make XMLHttpRequestProgressEventThrottle a weak pointer.
Instead, we take a weak pointer of its weak target and inform its weak target that the timer fired.

No observable change of behavior.

* Source/WebCore/xml/XMLHttpRequest.h:
* Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.cpp:
(WebCore::XMLHttpRequestProgressEventThrottle::updateProgress):
(WebCore::XMLHttpRequestProgressEventThrottle::dispatchThrottledProgressEventIfNeeded):
(WebCore::XMLHttpRequestProgressEventThrottle::dispatchThrottledProgressEventTimerFired): Deleted.
* Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.h:

Canonical link: <a href="https://commits.webkit.org/297934@main">https://commits.webkit.org/297934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58a5d8ebe5bafbc6daba55a0fe7c878ade8bbc48

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113496 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23649 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119677 "Built successfully") | [💥 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64262 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115393 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41778 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86351 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/64262 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116444 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27013 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102031 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66679 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26280 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20151 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63403 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20229 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122911 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30267 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95211 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40901 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98235 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94957 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24225 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40079 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17892 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36673 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40394 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45895 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40052 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43365 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41809 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (approved):
Looks good, but I think you meant to say that we no longer have to make it a WeakPtr in the commit message. Not that the current state of things is wrong.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fd1dd2630c03875819647c227b5d6d8a7b1d1798. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47886

youennf commented on Source/WebCore/xml/XMLHttpRequest.h:
```diff
@@ -139,6 +139,8 @@ class XMLHttpRequest final : public ActiveDOMObject, public RefCounted<XMLHttpRe
     using EventTarget::dispatchEvent;
     void dispatchEvent(Event&) override;
 
+    void dispatchThrottledProgressEventIfNeeded() { m_progressEventThrottle.dispatchThrottledProgressEventIfNeeded(); }
```
I hesitated between this approach and the patch's approach. I went with the patch's approach as it removes the need for `XMLHttpRequestProgressEventThrottle` to be CanMakeWeakPtr.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0ed5e986529357c9fdb26cfd01f688fbcfdf6172. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47886

annevk commented on Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.cpp:
```diff
@@ -68,10 +68,10 @@ void XMLHttpRequestProgressEventThrottle::updateProgress(bool isAsync, bool leng
-                    weakThis->dispatchThrottledProgressEventTimerFired();
-            });
+            minimumProgressEventDispatchingInterval, minimumProgressEventDispatchingInterval, TaskSource::Networking, [weakTarget = WeakPtr { m_target.get() }] {
+            if (RefPtr protectedTarget = weakTarget.get())
+                protectedTarget->dispatchThrottledProgressEventIfNeeded();
```
```suggestion
            if (RefPtr target = weakTarget.get())
                target->dispatchThrottledProgressEventIfNeeded();
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/58a5d8ebe5bafbc6daba55a0fe7c878ade8bbc48. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47886

youennf committed:
XMLHttpRequestProgressEventThrottle can be a weak pointer without being refable
rdar://155610120
https://bugs.webkit.org/show_bug.cgi?id=295774

Reviewed by Anne Van Kesteren.

We no longer need to make XMLHttpRequestProgressEventThrottle a weak pointer.
Instead, we take a weak pointer of its weak target and inform its weak target that the timer fired.

No observable change of behavior.

* Source/WebCore/xml/XMLHttpRequest.h:
* Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.cpp:
(WebCore::XMLHttpRequestProgressEventThrottle::updateProgress):
(WebCore::XMLHttpRequestProgressEventThrottle::dispatchThrottledProgressEventIfNeeded):
(WebCore::XMLHttpRequestProgressEventThrottle::dispatchThrottledProgressEventTimerFired): Deleted.
* Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.h:

Canonical link: https://commits.webkit.org/297934@main
